packages/vm/src/context.rs

```diff
@@ -62,229 +58,243 @@ impl GasState {
     }
 }
 
-struct ContextData<S: Storage, Q: Querier> {
-    gas_state: GasState,
-    storage: Option<S>,
-    storage_readonly: bool,
-    querier: Option<Q>,
-    /// A non-owning link to the wasmer instance
-    wasmer_instance: Option<NonNull<WasmerInstance>>,
+/// A ContextEnv is an env that provides access to the
+/// ContextData. The env is clonable but a clone accesses
+/// the same underlying data.
+pub struct Env<S: Storage, Q: Querier> {
+    data: Arc<RwLock<ContextData<S, Q>>>,
 }
 
-pub fn setup_context<S: Storage, Q: Querier>(gas_limit: u64) -> (*mut c_void, fn(*mut c_void)) {
-    (
-        create_unmanaged_context_data::<S, Q>(gas_limit),
-        destroy_unmanaged_context_data::<S, Q>,
-    )
+impl<S: Storage, Q: Querier> Clone for Env<S, Q> {
```

Comment:
A bit surprised that cannot be auto-derived (seems cleaner with derive statement), but if the compiler doesn't do it, fair enough.